File: android-compiler/src/main/java/com/pranav/lib_android/task/java/D8Task.java
package com.pranav.lib_android.task.java;

import com.android.tools.r8.D8;
import com.android.tools.r8.D8Command;
import com.android.tools.r8.OutputMode;

import com.pranav.common.util.ConcurrentUtil;
import com.pranav.common.util.FileUtil;
import com.pranav.lib_android.interfaces.*;

import java.io.File;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class D8Task extends Task {

    private Exception ex = null;

    @Override
    public void doFullTask() throws Exception {
        ConcurrentUtil.execute(
                () -> {
                    try {
                        D8.run(
                                D8Command.builder()
                                        .setOutput(
                                                Paths.get(FileUtil.getBinDir()),
                                                OutputMode.DexIndexed)
                                        .addLibraryFiles(
                                                Paths.get(
                                                        FileUtil.getClasspathDir(), "android.jar"))
                                        .addProgramFiles(
                                                getClassFiles(
                                                        new File(FileUtil.getBinDir(), "classes")))
                                        .build());

                    } catch (Exception e) {
                        ex = e;
                    }
                });
        if (ex != null) {
            throw ex;
        }
    }

    private ArrayList<Path> getClassFiles(File root) {
        var paths = new ArrayList<Path>();

        var files = root.listFiles();
        if (files != null) {
            for (var f : files) {
                if (f.isFile()) {
                    paths.add(f.toPath());
                } else {
                    paths.addAll(getClassFiles(f));
                }
            }
        }
        return paths;
    }

    @Override
    public String getTaskName() {
        return "D8 Task";
    }
}

File: android-compiler/src/main/java/org/cosmic/ide/android/task/dex/D8Task.kt
package org.cosmic.ide.android.task.dex

import com.android.tools.r8.CompilationMode
import com.android.tools.r8.D8
import com.android.tools.r8.D8Command
import com.android.tools.r8.OutputMode
import com.android.tools.r8.DexIndexedConsumer.DirectoryConsumer
import org.cosmic.ide.CompilerUtil
import org.cosmic.ide.android.interfaces.Task
import org.cosmic.ide.project.Project
import java.io.File
import java.nio.file.Path
import java.nio.file.Paths

class D8Task : Task {

    /*
     * Compile a jar file to a Dalvik Executable (Dex) File.
     *
     * @param jarFile the jar file to compile
     */
    fun compileJars(jarFiles: List<Path>, dir: Path) {
        D8.run(
            D8Command.builder()
                .setMinApiLevel(26)
                .setMode(CompilationMode.DEBUG)
                .addClasspathFiles(CompilerUtil.platformPaths)
                .addProgramFiles(jarFiles)
                .setOutput(dir, OutputMode.DexIndexed)
                .setProgramConsumer(
                    DirectoryConsumer(dir)
                )
                .build()
        )
    }

    /*
     * Compile class files of project to a Dalvik Executable (Dex) File.
     *
     * @param project the project to compile.
     */
    @Throws(Exception::class)
    override fun doFullTask(project: Project) {
        D8.run(
            D8Command.builder()
                .setMinApiLevel(26)
                .setMode(CompilationMode.DEBUG)
                .addClasspathFiles(CompilerUtil.platformPaths)
                .addProgramFiles(
                    getClassFiles(File(project.binDirPath, "classes"))
                )
                .setOutput(Paths.get(project.binDirPath), OutputMode.DexIndexed)
                .build()
        )

        // Compile libraries
        val folder = File(project.libDirPath)
        if (folder.exists() && folder.isDirectory) {
            val libs = folder.listFiles()
            if (libs != null) {
                val toDex = mutableListOf<Path>()
                val libDexes = File(project.buildDirPath, "libs")
                libDexes.mkdirs()
                // Check if all libs have been pre-dexed or not
                for (lib in libs) {
                    val outDex = File(libDexes, lib.nameWithoutExtension + ".dex")

                    if (lib.extension == "jar" && !outDex.exists()) {
                        toDex.add(lib.toPath())
                    }
                }
                compileJars(toDex, Paths.get(project.buildDirPath, "libs"))
            }
        }
    }

    /*
     * Find all classes recursively in a directory.
     *
     * @param root directory to search in.
     */
    private fun getClassFiles(root: File): List<Path> {
        val paths = mutableListOf<Path>()

        root.walk().forEach {
            if (it.extension == "class") {
                paths.add(it.toPath())
            }
        }
        return paths
    }
}
